### DISCUSSION TITLE
`sea-orm-cli generate entity` creates `mod.rs` in library crate

### DISCUSSION THREAD

**Discussion by viktorbahr** [SeaQL/sea-orm#604] [Q&A]
## Preface
Since I'm relatively new to Rust, I'm not 100% sure if I'm missing something here. In case I did, please excuse my ignorance!

## Description

When pointing the output directory of `sea-orm-cli generate entity` to the `src` directory of an `entity` library crate, it generates a (what I assume to be an unneeded) `mod.rs` file while leaving the `lib.rs` untouched. Using the generated entity from another crate in the workspace (i.e. the `migration` crate) then requires to manually copy the changes made to the `mod.rs` file over to the `lib.rs` file of the `entity` crate. 

## Steps to Reproduce

Requirements: Database access and valid `DB_URL` env set

1. Create a new crate.
2. Inside that crate, create an empty `entity` library crate as described in the [migration docs](https://www.sea-ql.org/SeaORM/docs/migration/setting-up-migration#entity-crate)
3. Additionally, create a new `migration` crate using `sea-orm-cli migrate init`
4. Add a migration to the `migration` crate that creates a new table in the DB
5. Apply the migration using `sea-orm-cli migrate up`
6. Generate the entity using `sea-orm-cli generate entity -o entity/src`

### Expected Behavior

The CLI should add the `pub mod` declaration for the newly generated entity (and `prelude`, `seaql_migrations`) to `entity/src/lib.rs`.

### Actual Behavior
The CLI generates a `mod.rs` file, containing the `pub mod` declarations.

## Versions

```
│   └── sea-orm v0.6.0
│       ├── sea-orm-macros v0.6.0 (proc-macro)
│       ├── sea-query v0.21.0
│       │   ├── sea-query-derive v0.2.0 (proc-macro)
│       ├── sea-strum v0.23.0
│       │   └── sea-strum_macros v0.23.0 (proc-macro)
│   └── sea-schema v0.5.1
│       ├── sea-orm v0.6.0 (*)
│       ├── sea-query v0.21.0 (*)
│       ├── sea-schema-derive v0.1.0 (proc-macro)
├── sea-orm v0.6.0 (*)
```

## Additional Information

To me it looks like that this is a mismatch between the behavior of the CLI and the structure proposed in the docs. To fix this, I'd propose adding a check for an existing `lib.rs` to the CLI, which, if triggered, would cause the changes that would otherwise be written to `mod.rs` to be made there. I could try to setup a PR for this in case my assumptions are validated by someone from the maintainer team.

**Comment by tyt2y3** [comment]
I think you can put them in a sub directory `/entities/mod.rs` and add the following to your lib.rs

```rust
mod entities;
pub use entities::*;
```

**Reply by tyt2y3** [reply]
> I'd propose adding a check for an existing lib.rs to the CLI, which, if triggered, would cause the changes that would otherwise be written to mod.rs to be made there.

@billy1624 your thought on this?

**Reply by billy1624** [reply]
Hey @viktorbahr, thanks for the suggestions!

> > I'd propose adding a check for an existing lib.rs to the CLI, which, if triggered, would cause the changes that would otherwise be written to mod.rs to be made there.
> 
> @billy1624 your thought on this?

I think we could add a `--lib` flag to CLI. With this flag enabled the content of `mod.rs` will be written to `lib.rs` instead. I want to make the behaviour explicit and avoid confusion.

How is that sounds? @viktorbahr

**Reply by viktorbahr** [reply]
Sounds good @billy1624 !
I could try to get a PR out for this next weekend.

**Reply by billy1624** [reply]
Thanks!! @viktorbahr Looking forward to it :)

**Reply by Samyak2** [reply]
I came across this issue too. I was using the directory structure mentioned [here](https://www.sea-ql.org/SeaORM/docs/migration/setting-up-migration#entity-crate). The note there says:

---

Finally, [generate SeaORM entities](https://www.sea-ql.org/SeaORM/docs/generate-entity/sea-orm-cli) with sea-orm-cli and **output the entity files to `entity/src` folder.**

---

As a temporary solution, I think this note can be changed to say:

output the entity files to `entity/src/entities` folder and add the following to `lib.rs`:
```rust
mod entities;
pub use entities::*;
```

---

Once the `--lib` flag is added, it would be nice to update the note to mention that this flag needs to be enabled while generating entities.

**Reply by billy1624** [reply]
Yeah, we can direct users to add that. See https://github.com/SeaQL/seaql.github.io/commit/b6c6c2438f94b82aa96b0998241b78f6e101a8e1

**Reply by Samyak2** [reply]
@billy1624 `entities` is not mentioned anywhere in that page, so it will confuse users as they wouldn't know what it refers to and what the contents of `lib.rs` should be after adding the entities.

Also, it would be better to add this information in the "If you don't have SeaORM entities defined?" note here: https://www.sea-ql.org/SeaORM/docs/migration/setting-up-migration/#entity-crate

**Reply by billy1624** [reply]
Hey @Samyak2, that's even better! https://github.com/SeaQL/seaql.github.io/commit/d8236e3df81692e3510be3b7b868bcd8ba27eb1e

**Reply by Samyak2** [reply]
@billy1624 that's nice! Although, the line right above it says ```output the entity files to `entity/src` folder.```. It should be changed to ```output the entity files to `entity/src/entities` folder.```

**Reply by billy1624** [reply]
Oh my. Nice catch! @Samyak2
https://github.com/SeaQL/seaql.github.io/commit/598a506feaeb811e53f08fa1f6bb6e06c2de2be2

**Reply by HigherOrderLogic** [reply]
> Sounds good @billy1624 ! I could try to get a PR out for this next weekend.

Hey @viktorbahr, are you still working on this? If not, I can take it over. Thanks!

**Reply by billy1624** [reply]
Hey everyone, I just open an issue to track the progress
- https://github.com/SeaQL/sea-orm/issues/934
